README.md

```diff
@@ -1,16 +1,268 @@
+JavaScript console when the code is executed?
+
+```js
+myFunc();
+// LOG: undefined
+// => undefined
+```
+
+It prints out `undefined`. What the heck?!
+
+You see, in JavaScript, hoisting only applies to variable _declarations_; not
+variable _assignments_. As a quick refresher on that terminology:
+
+```js
+// Declaration:
+let hello;
+
+// Assignment:
+hello = 'World!';
+
+// Declaration and assignment on the same line:
+let goodnight = 'Moon';
+```
+
+During the compilation phase, the JavaScript engine initializes the variable
+`hello`, storing it in memory. At this point, however, **no value is assigned to
+the variable**. As far as the JavaScript engine is concerned, the variable
+`hello` exists, but it contains `undefined`.
+
+The variable will contain `undefined` until it's assigned a different value
+during the execution phase. Because of this odd behavior, you'll often see
+variable hoisting explained by taking some sample code...
+
+```js
+function myFunc () {
+  console.log(hello);
+
+  var hello = 'World!';
+}
+```
+
+and rearranging it to better indicate the order of events:
+
+```js
+function myFunc () {
+  var hello;
+
+  console.log(hello);
+
+  hello = 'World!';
+}
+```
+
+When rearranged, it's clear that the variable is initialized as `undefined`,
+that it still contains `undefined` when it's logged out to the console, and that
+only after the logging event is it assigned the value of `'World!'`. However,
+armed with knowledge of what's going on under the surface (the distinct
+compilation and execution phases), we don't need to move the code around. When
+we invoke the following function, five things happen:
+
+```js
+function myFunc () {
+  console.log(hello);
+
+  var hello = 'World!';
+
+  return hello;
+}
+
+myFunc();
+// LOG: undefined
+// => "World!"
+```
+
+1. The declaration of `hello` (`var hello`) is evaluated during the compilation
+phase, and the identifier, `hello`, is stored in memory as `undefined`.
+
+2. The execution phase starts, and the JavaScript engine begins stepping through
+the code, executing each line in turn.
+
+3. At the first line, `console.log(hello);`, the value of `hello` is still
+`undefined`, and that's exactly what gets logged out to the console.
+
+4. At the second line, the value `'World!'` is assigned to the variable `hello`.
+From this point on, all references to `hello` in this scope will evaluate to
+`'World!'`.
+
+5. At the final line, we `return` the value of `hello`, which by now has been
+assigned and evaluates to `'World!'`.
+
+### Avoiding the Confusion of `var` Hoisting
+
+There are two ways to keep the JavaScript engine from 'hoisting' your variables:
+
+1. If, for whatever reason, your current project requires that you use `var`,
+follow our rule for function declarations and declare everything at the **top**
+of its scope. For example, if you need to declare a variable within a function,
+declare it at the **top** of that function:
+
+```js
+// BAD
```

Comment:
We've been avoiding the use of "BAD/GOOD" examples in other READMEs. I think we used "Recommended/Not Recommended" or something to that nature. I can double-check in S6 where I think we had a couple examples of this.